test/x509aux.c

```diff
@@ -27,7 +27,7 @@ static int test_certs(int num)
     unsigned char *data = 0;
     long len;
     typedef X509 *(*d2i_X509_t)(X509 **, const unsigned char **, long);
-    typedef int (*i2d_X509_t)(X509 *, unsigned char **);
+    typedef int (*i2d_X509_t)(const X509 *, unsigned char **);
```

Comment:
Ouch. This can actually be seen as an API change, if something in our public header files forced you to make this change...

Reply:
Yes, you are right. The compiler will complain unless the typedef is changed.

**BUT:** _If the possibility of signature changes in typedef's would be an impediment for backporting, then all our backported constifications would be invalid!_

Because the user could have had typedef'd any of those constified functions.

Since no one ever seems to complain when we backport constifications, I guess that it's rather uncommon practice. Please also keep in mind that we are looking at some testing code which digs deep into the d2i/i2d code. That's probably also rather uncommon practice for user applications.

Reply:
I did not check the policies whether it is written down in our policies, but common practice has always been to consider constifications in the argument list a compatible API change which can be backported. I'm not sure whether we really want to change that policy, because functions which are not const-correct really suck.

Reply:
With compatible, we do mean that it won't break someone's application build when they are updating their openssl installation (going from 1.1.1c to 1.1.1d, for example). If something we do with our public API breaks its normal use, we've done wrong.

Our policy is clear on the matter:

> No API or ABI breaking changes are allowed in a minor or patch release. 

Ref: https://www.openssl.org/policies/releasestrat.html

Reply:
> With compatible, we do mean that it won't break someone's application build when they are updating their openssl installation (going from 1.1.1c to 1.1.1d, for example). If something we do with our public API breaks its normal use, we've done wrong.
> 
> Our policy is clear on the matter:
> 
> > No API or ABI breaking changes are allowed in a minor or patch release.
> 
> Ref: https://www.openssl.org/policies/releasestrat.html


As far as I understood it we always considered constifications to be security enhancements and as such acceptable for the stable branch. There are five constifications which were backported to OpenSSL_1_1_1-stable, two of which were approved by you (marked with a dot).

* dc9f2a69f3 issue-8973: Added const to parameters for values that were not altered
  2c5afadc86 Change SSL parameter SSL_session_reused const
  edfd50d8db Make OCSP_id_cmp and OCSP_id_issuer_cmp accept const params
  00f2bafec1 Make some simple getters take const SSL/SSL_CTX
* 92791b972c EVP_PKEY_size declared to take a const parameter

Are you saying that _nobody_ in the team was aware that changing a function signature by adding a 'const' to an input parameter is an API change? And that there might be a tiny risk that some compiler may complain somewhere?

I'd  rather believe that everybody accepted the tiny risk of breaking some build out there in favor of the security benefits they introduce. Because they are able to catch programming errors which result in invalid memory write access. So they were treated like bugfixes. 

I still think that backporting constifications is good practice and we should keep it up. The Release Strategy should not prevent us from doing it, so the wording might need to clarified in that point by an @openssl/omc vote.

Reply:
What happens if the address of a function is taken and one of its arguments changes from a pointer to a pointer to const?

Reply:
You'll get an `-Werror=incompatible-pointer-types` error in gcc:

```make
test/x509aux.c: In function 'test_certs':
test/x509aux.c:42:26: error: initialization of 'i2d_X509_t' {aka 'int (*)(struct x509_st *, unsigned char **)'} from incompatible pointer type 'int (*)(const X509 *, unsigned char **)' {aka 'int (*)(const struct x509_st *, unsigned char **)'} [-Werror=incompatible-pointer-types]
         i2d_X509_t i2d = trusted ? i2d_X509_AUX : i2d_X509;
                          ^~~~~~~
```

when I revert the change in x509aux.change

```diff
diff --git a/test/x509aux.c b/test/x509aux.c
index 2b30c770c5..e41f1f6809 100644
--- a/test/x509aux.c
+++ b/test/x509aux.c
@@ -27,7 +27,7 @@ static int test_certs(int num)
     unsigned char *data = 0;
     long len;
     typedef X509 *(*d2i_X509_t)(X509 **, const unsigned char **, long);
-    typedef int (*i2d_X509_t)(const X509 *, unsigned char **);
+    typedef int (*i2d_X509_t)(X509 *, unsigned char **);
     int err = 0;
     BIO *fp = BIO_new_file(test_get_argument(num), "r");
     X509 *reuse = NULL;
```

Reply:
Note that it's just a warning.  But we treat warnings as errors.

Reply:
Ah, hold on, I misread the code, I thought that one of our callback signatures had changed, but that is not the case.  And otherwise, we have already accepted constification of our public functions, so this PR should be acceptable as well under the same terms.

However, what @paulidale says raises a good point that I have simply not thought of before.  Personally, I *hope* that people write wrappers and take the address of those, but one never knows...  however, if they do use the address of our functions directly, their builds may break because of constification.  Is that acceptable or not?  If not, all those constifying commits you mention, @mspncp, are invalid and should be reverted.  I hope we don't go there.  This is certainly something for @openssl/omc to discuss.

Reply:
(some of those commits are probably already released...  if no one has complained, perhaps we can get away with it?)

Reply:
**Correction**:  the build failed only for me because I configured `--strict-warnings`

```
perl Configure debug-linux-x86_64 --strict-warnings
```


When I use the default configuration

```		 
~/src/openssl$ ./config
Operating system: x86_64-whatever-linux2
```

then the build succeeds.

I hope this fact makes decisions simpler.

Reply:
Agreed, an OMC discussion is in order.  We have accepted _const_ changes before and I thought it was the agreed opinion that these were okay.

That compilers are dumb enough to not realise that a _constified_ function **is** compatible isn't ideal.

Reply:
> However, what @paulidale says raises a good point that I have simply not thought of before. 

Huh...? I thought this was the point of our discussion from the very beginning?

Reply:
> That compilers are dumb enough to not realise that a constified function is compatible isn't ideal.

Compilers do realize it, that's why it's just a warning, not an error. I know about implicit conversions which are not allowed on purpose, because it would allow circumventing the constness constraint in subtle ways. Maybe there is a similar reason in our case?  (cc @bernd-edlinger)

Reply:
Yes, you get a warning if you assign the function pointer no matter how small the difference is.
When I wrote the gcc warning -Wcast-function-type it was initially more strict than it is now.
But that turned out to be contra-productive, since if the parameters are very similar,
mybe different only in const/volatile then the explict cast should be taken as indication
that the user wanted to silence a warning if the parameters are very similar as in this case.
OTOH, if someone uses -Werror and no type cast in the assignment, then it
will break due to this change.

Reply:
note, this issue here has some similarity to #9296

Reply:
Thanks for pointing out #9296. BTW: Would the compiler treat a constness mismatch in the signature of the callback as warning or as error? (I could ask the compiler, but I'm lazy and hope you can tell me right away.)

Reply:
it depends.
It is a warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
but it can be an error if -Werror is used,
and is usually an error in C++ (unless -fpermissive is used)

Reply:
So, in case someone passes a pointer to one of our functions that have been constified, they may get this kind of warning by default (which is reasonable, since they are thereby un-constifying it).

I would say that building an application with `-Werror`, while good for safety of one's own programming practices, is "asking for it" when applying that to code from external libraries.  In my personal view, that makes it an acceptable.  However, it would be good if the rest of @openssl/omc gave their opinion.